[controller-manager] Switch `ManagedSeedSet` controller to `controller-runtime` (#6742)
gardener/gardener: docs/concepts/controller-manager.md

Consequently, to ensure that `ExposureClass`s in-use are always present in the system until the last referring `Shoot` gets deleted, the controller adds a finalizer which is only released when there is no `Shoot` referencing the `ExposureClass` anymore.

### [`ManagedSeedSet` Controller](../../pkg/controllermanager/controller/managedseedset)

> `ManagedSeedSet` objects maintain a stable set of replica of `ManagedSeed`s, i.e they guarantee the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.

Reviewer: ```suggestion
`ManagedSeedSet` objects maintain a stable set of replicas of `ManagedSeed`s, i.e they guarantee the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.
```

`ManagedSeedSet` controller creates and deletes `ManagedSeed`s and `Shoot`s in response to changes to the replicas and selector fields. For more info refer to the [`ManagedSeedSet` proposal document](../../docs/proposals/13-automated-seed-management.md#managedseedsets).

> 1. The reconciler first gets all the replicas of the given `ManagedSeedSet` in the `ManagedSeedSet`'s namespace and with the matching selector. Each replica is a tuple that contains a `ManagedSeed`, its corresponding `Seed` and `Shoot` objects.

Reviewer: What do you mean by "each replica is a tuple"? Maybe remove this or use "group" or "structure" instead?

1. Then the pending replica is retrieved, if it exists.
1. Next it determines the ready, postponed and deletable replica.
    - A replica is considered `ready` when a `Seed` owned by a `ManagedSeed` has been registered either directly or by deploying `gardenlet` into a `Shoot`, the `Seed` is `Ready` and the `Shoot`'s status is `Healthy`.

>     - If a replica is not ready and it is not pending, i.e it is not specified in the `ManagedSeed`'s `status.pendingReplica` field, then it is added to the `postponed` replicas.

Reviewer: ```suggestion
    - If a replica is not ready and it is not pending, i.e. it is not specified in the `ManagedSeed`'s `status.pendingReplica` field, then it is added to the `postponed` replicas.
```

>     - A replica is deletable if it has no scheduled `Shoot`s and is not protected by the `protect-from-deletion` annotation.

Reviewer: > by the `protect-from-deletion` annotation.

What's that? Is that something specific to this controller? Maybe introduce this somewhere above.

Author: Yes, a replica is marked as deletable when shoot and managedseed of the replica does not have `seedmanagement.gardener.cloud/protect-from-deletion` annotation present and if there is no shoot scheduled for the replica.

> 1. Finally, it checks the actual and target replica counts. If actual count is less than the target count, the controller scales up the replicas by creating a new replica to match the desired target count. And if the actual count is more than the target, the controller deletes a replica to match the desired count. Before `scaleOut` or `scaleIn`, the controller first reconciles the pending replica and makes sure the replica is ready before moving on the next replica.

Reviewer: ```suggestion
1. Finally, it checks the actual and target replica counts. If actual count is less than the target count, the controller scales up the replicas by creating new replicas to match the desired target count. If the actual count is more than the target, the controller deletes replicas to match the desired count. Before scale-out or scale-in, the controller first reconciles the pending replica (there can always only be one) and makes sure the replica is ready before moving on the next one.
```

>     * `Scaleout`(actual count < target count)

Reviewer: ```suggestion
    * Scale-out (actual count < target count)
```

>         - During the `scaleOut` phase the controller first creates the `Shoot` object from the `ManagedSeedSet`'s `spec.shootTemplate` field and adds the replica to the `status.pendingReplica` of the `ManagedSeedSet`.

Reviewer: ```suggestion
        - During the scale-out phase, the controller first creates the `Shoot` object from the `ManagedSeedSet`'s `spec.shootTemplate` field and adds the replica to the `status.pendingReplica` of the `ManagedSeedSet`.
```

>         - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica or step. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed controller` and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.

Reviewer 2: ```suggestion
        - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica or step. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed` controller and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.

>         - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica or step. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed` controller and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.

Reviewer: ```suggestion
        - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed` controller and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.
```

>     * `ScaleIn`(actual count > target count)

Reviewer: ```suggestion
    * Scale-in (actual count > target count)
```

>         - During `scaleIn` phase the controller first determines the replica that can be deleted. From the deletable replicas it chooses the one with lowest priority and deletes it.

Reviewer: ```suggestion
        - During scale-in phase, the controller first determines the replica that can be deleted. From the deletable replicas it chooses the one with lowest priority and deletes it.
```

Reviewer: Which one gets the lowest priority?

> 

Reviewer: ```suggestion
```

Consequently, to ensure that `ExposureClass`s in-use are always present in the system until the last referring `Shoot` gets deleted, the controller adds a finalizer which is only released when there is no `Shoot` referencing the `ExposureClass` anymore.

### [`ManagedSeedSet` Controller](../../pkg/controllermanager/controller/managedseedset)

> `ManagedSeedSet` object maintains a stable set of replica of `ManagedSeed`s i.e it gurantee the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.

Reviewer 2: ```suggestion
`ManagedSeedSet` object maintains a stable set of replica of `ManagedSeed`s i.e it gurantees the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.
```

Reviewer: It's rather `guarantees`, I think :)

> `ManagedSeedSet` object maintains a stable set of replica of `ManagedSeed`s i.e it gurantee the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.

Reviewer: ```suggestion
`ManagedSeedSet` objects maintain a stable set of replica of `ManagedSeed`s, i.e they guarantee the availability of a specified number of identical `ManagedSeed`s on an equal number of identical `Shoot`s.
```

> `ManagedSeedSet` controller creates and deletes `ManagedSeed`s and `Shoot`s in response to changes to the replicas and selector fields. For more info on `ManagedSeedSet` resource [refer](../../docs/proposals/13-automated-seed-management.md#managedseedsets).

Reviewer 2: ```suggestion
`ManagedSeedSet` controller creates and deletes `ManagedSeed`s and `Shoot`s in response to changes to the replicas and selector fields. For more info refer to the [`ManagedSeedSet` proposal document](../../docs/proposals/13-automated-seed-management.md#managedseedsets).
```

> #### "Actuator"

Reviewer 2: If we will mention that this reconciler uses an actuator, perhaps it makes sense to mention that there is an `Actuator` interface for which the default implementation does the steps described below. Similar to https://github.com/gardener/gardener/blob/master/docs/extensions/infrastructure.md#actuator-interface.
Although, this is more-or-less an implementation detail, so not sure if the `Actuator` struct/interface should be mentioned here and perhaps only leave the steps listed below?

Reviewer: I would not mention the actuator explicitly. This is an implementation detail and is not needed to be documented here. Let's only talk about the "reconciler", i.e. let's probably remove this headline.

> 1. The actuator first gets all the existing replicas of the given set i.e it gets managed seeds, shoots and seeds in the set's namespace and with the matching selector.

Reviewer 2: ```suggestion
1. The actuator first gets all the replicas of the given `ManagedSeedSet` in the `ManagedSeedSet`'s namespace and with the matching selector. Each replica is a tuple that contains a `ManagedSeed`, its corresponding `Seed` and `Shoot` objects.
```

> 1. Get the pending replica, if any.

Reviewer 2: ```suggestion
1. Then the pending replica is retrieved, if it exists.
```

> 1. Determine ready, postponed and deletable replica.

Reviewer 2: ```suggestion
1. Next it determines the ready, postponed and deletable replica.
```

>     - A replica is considered `ready` when `ManagedSeed`'s seed has been registered either directly or by deploying gardenlet into the shoot, `Seed` is ready and `Shoot`'s status is `heathy`.

Reviewer 2: ```suggestion
    - A replica is considered `ready` when a `Seed` owned by a `ManagedSeed` has been registered either directly or by deploying `gardenlet` into a `Shoot`, the `Seed` is `Ready` and the `Shoot`'s status is `Healthy`.
```

>     - If the replica is not ready and it is not in the pending status then it added to the `postponed` replicas.

Reviewer 2: ```suggestion
    - If a replica is not ready and it is not pending, i.e it is not specified in the `ManagedSeed`'s `status.pendingReplica` field, then it is added to the `postponed` replicas.
```

>     - A replica is deletable if it has no scheduled shoots and is not protected by the "protect-from-deletion" annotation.

Reviewer 2: ```suggestion
    - A replica is deletable if it has no scheduled `Shoot`s and is not protected by the "protect-from-deletion" annotation.
```

> 1. Next, it checks the actual and target replica counts. If actual count is less than the target count, the controller scales up the replicas by creating new replica to match the desired target count. And if the actual count is more than the target, the controller deletes replica to match the desired count. Before `scaleOut` or `scaleIn`, the controller first reconciles the pending replica and makes sure the replica is ready before moving on the next replica.

Reviewer 2: ```suggestion
1. Finally, it checks the actual and target replica counts. If actual count is less than the target count, the controller scales up the replicas by creating a new replica to match the desired target count. And if the actual count is more than the target, the controller deletes a replica to match the desired count. Before `scaleOut` or `scaleIn`, the controller first reconciles the pending replica and makes sure the replica is ready before moving on the next replica.
```

    * `Scaleout`(actual count < target count)

>         - During `scaleOut` phase the controller first creates the `Shoot` object from `ManagedSeedSet`'s shoot templates and adds the replica to the `pendingReplica` of `ManagedSeedSet`.

Reviewer 2: ```suggestion
        - During the `scaleOut` phase the controller first creates the `Shoot` object from the `ManagedSeedSet`'s `spec.shootTemplate` field and adds the replica to the `status.pendingReplica` of the `ManagedSeedSet`.
```

>         - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready then only it moves to the next replica or step. One the create shoot is reconciled successfully, `ManagedSeed` object is created from the `ManagedSeedSet`'s managedseed template. ManagedSeed object is taken care by `ManagedSeed controller`. Once the replica's `Seed` becomes ready and `Shoot` becomes healthy, the replica becomes ready.

Reviewer 2: ```suggestion
        - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica or step. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed controller` and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.
```

`ManagedSeedSet` controller creates and deletes `ManagedSeed`s and `Shoot`s in response to changes to the replicas and selector fields. For more info refer to the [`ManagedSeedSet` proposal document](../../docs/proposals/13-automated-seed-management.md#managedseedsets).

1. The reconciler first gets all the replicas of the given `ManagedSeedSet` in the `ManagedSeedSet`'s namespace and with the matching selector. Each replica is a struct that contains a `ManagedSeed`, its corresponding `Seed` and `Shoot` objects.
1. Then the pending replica is retrieved, if it exists.
1. Next it determines the ready, postponed and deletable replica.
    - A replica is considered `ready` when a `Seed` owned by a `ManagedSeed` has been registered either directly or by deploying `gardenlet` into a `Shoot`, the `Seed` is `Ready` and the `Shoot`'s status is `Healthy`.
    - If a replica is not ready and it is not pending, i.e. it is not specified in the `ManagedSeed`'s `status.pendingReplica` field, then it is added to the `postponed` replicas.

>     - A replica is deletable if it has no scheduled `Shoot`s and replica's Shoot and ManagedSeed does not have `seedmanagement.gardener.cloud/protect-from-deletion` annotation.

Reviewer: ```suggestion
    - A replica is deletable if it has no scheduled `Shoot`s and replica's `Shoot` and `ManagedSeed` do not have the `seedmanagement.gardener.cloud/protect-from-deletion` annotation.
```

1. Finally, it checks the actual and target replica counts. If actual count is less than the target count, the controller scales up the replicas by creating new replicas to match the desired target count. If the actual count is more than the target, the controller deletes replicas to match the desired count. Before scale-out or scale-in, the controller first reconciles the pending replica (there can always only be one) and makes sure the replica is ready before moving on the next one.
    * `Scale-out`(actual count < target count)

>         - During the scale-out phase the controller first creates the `Shoot` object from the `ManagedSeedSet`'s `spec.shootTemplate` field and adds the replica to the `status.pendingReplica` of the `ManagedSeedSet`.

Reviewer: ```suggestion
        - During the scale-out phase, the controller first creates the `Shoot` object from the `ManagedSeedSet`'s `spec.shootTemplate` field and adds the replica to the `status.pendingReplica` of the `ManagedSeedSet`.
```

        - For the subsequent reconciliation steps, the controller makes sure that the pending replica is ready before proceeding to the next replica. Once the `Shoot` is created successfully, the `ManagedSeed` object is created from the `ManagedSeedSet`'s `spec.template`. The `ManagedSeed` object is reconciled by the `ManagedSeed` controller and a `Seed` object is created for the replica. Once the replica's `Seed` becomes ready and the `Shoot` becomes healthy, the replica also becomes ready.
    * `Scale-in`(actual count > target count)

>         - During scale-in phase the controller first determines the replica that can be deleted. From the deletable replicas it chooses the one with lowest priority and deletes it. Priority is determined in following order:

Reviewer: ```suggestion
        - During scale-in phase, the controller first determines the replica that can be deleted. From the deletable replicas it chooses the one with lowest priority and deletes it. Priority is determined in following order:
```

>             - First compare replica statuses. Replicas with "less advanced" status are considered lower priority. For example, replica with StatusShootReconciling status has lower value 3 than replica with StatusShootReconciled status which has value as 5. Hence in this case replica with StatusShootReconciling status will have lower priority and will be considered for deletion.

Reviewer: ```suggestion
            - First compare replica statuses. Replicas with "less advanced" status are considered lower priority. For example, replica with `StatusShootReconciling` status has a lower value than replica with `StatusShootReconciled` status. Hence, in this case replica with `StatusShootReconciling` status will have lower priority and will be considered for deletion.
```

>             - Then, compare replica seed readiness. Replicas with non-ready seeds are considered lower priority.

Reviewer: ```suggestion
            - Then, the replicas are compared with the readiness of their `Seed`s. Replicas with non-ready `Seed`s are considered lower priority.
```

>             - Then, compare replica shoot health statuses. Replicas with "worse" status are considered lower priority.

Reviewer: ```suggestion
            - Then, the replicas are compared with the health statuses of their `Shoot`s. Replicas with "worse" statuses are considered lower priority.
```

>             - Finally, compare replica ordinals. Replicas with lower ordinals are considered lower priority.

Reviewer: ```suggestion
            - Finally, the replica ordinals are compared. Replicas with lower ordinals are considered lower priority.
```
